### PR TITLE
feat: Integrate tracers and implement circuits tracer in vm2

### PR DESCRIPTION
**Pull request by montekki**
## What ❔

Integrates tracers support into the codebase and implements the circuits tracer.

## Why ❔

Tracers are required for some VM applications, e.g. to determine batch seal criteria and for tracing calls.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

**Review by slowli — COMMENTED** [review]
Is performance degradation on `iai` benches expected? (AFAIU, it *is* real degradation compared to the `main` branch, e.g. judging by [these results on `main`](https://github.com/matter-labs/zksync-era/actions/runs/10686743019/job/29622798327#step:5:1195).) Have you checked whether it translates to the real-time degradation?

File: core/lib/multivm/src/versions/vm_fast/circuits_tracer.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
+use circuit_sequencer_api_1_5_0::{geometry_config::get_geometry_config, toolset::GeometryConfig};
+use vm2::{CycleStats, Opcode, OpcodeType, StateInterface, Tracer};
+use zksync_vm_interface::CircuitStatistic;
+
+use crate::vm_latest::tracers::circuits_capacity::*;
+
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct CircuitsTracer {
```
[the remark this answers is not in the document]
Comment (joonazan) [reply core/lib/multivm/src/versions/vm_fast/circuits_tracer.rs:8]: Interesting, I did not know about pub(super).

File: etc/contracts-test-data/contracts/failed-call/failed_call.sol [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+contract FailedCall {
+    bool public success;
+    bytes1 public data_first_byte;
+    
+    constructor() {
+        address MSG_VALUE_SIMULATOR = 0x0000000000000000000000000000000000008009;
+
+		while (gasleft() > 20000) {
+            // Burn gas so that there's about `gasToPass` left before the external call.
```
[the remark this answers is not in the document]
Comment (joonazan) [reply etc/contracts-test-data/contracts/failed-call/failed_call.sol:13]: I don't know. Fedor wrote this.

**Comment by joonazan** [comment]
The slowdown is real but mostly under 10%. heap_read_write shows +15% on my M1 even though it shows no change in iai.
This can be investigated later when we want to really optimize performance.

**Review by joonazan — APPROVED** [review]

**Pull request merged**